ssl/quic/quic_fc.c

```diff
@@ -0,0 +1,366 @@
+
+    if (rxfc->parent == NULL)
+        return 0;
+
+    if (rxfc->is_fin && ((is_fin && rxfc->hwm != end) || end > rxfc->hwm)) {
+        /* Stream size cannot change after the stream is finished */
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
+
+    if (is_fin)
+        rxfc->is_fin = 1;
+
+    if (end > rxfc->hwm) {
+        delta = end - rxfc->hwm;
+        rxfc->hwm = end;
+
+        on_rx_controlled_bytes(rxfc, delta);           /* result ignored */
+        on_rx_controlled_bytes(rxfc->parent, delta);   /* result ignored */
+    } else if (end < rxfc->hwm && is_fin) {
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
+
+    return 1;
+}
+
+/* threshold = 3/4 */
+#define WINDOW_THRESHOLD_NUM 3
+#define WINDOW_THRESHOLD_DEN 4
+
+static int rxfc_cwm_bump_desired(QUIC_RXFC *rxfc)
+{
+    int err = 0;
+    uint64_t window_rem = rxfc->cwm - rxfc->rwm;
+    uint64_t threshold
+        = safe_mul_uint64_t(rxfc->cur_window_size,
+                            WINDOW_THRESHOLD_NUM, &err) / WINDOW_THRESHOLD_DEN;
+
+    if (err)
+        /*
+         * Extremely large window should never occur, but if it does, just use
+         * 1/2 as the threshold.
+         */
+        threshold = rxfc->cur_window_size / 2;
+
+    return window_rem <= threshold;
+}
+
+static int rxfc_should_bump_window_size(QUIC_RXFC *rxfc, OSSL_TIME rtt)
+{
+    /*
+     * dt: time since start of epoch
+     * b:  bytes of window consumed since start of epoch
+     * dw: proportion of window consumed since start of epoch
+     * T_window: time it will take to use up the entire window, based on dt, dw
+     * RTT: The current estimated RTT.
+     *
+     * b        = rwm - esrwm
+     * dw       = b / window_size
+     * T_window = dt / dw
+     * T_window = dt / (b / window_size)
+     * T_window = (dt * window_size) / b
+     *
+     * We bump the window size if T_window < 4 * RTT.
+     *
+     * We leave the division by b on the LHS to reduce the risk of overflowing
+     * our 64-bit nanosecond representation, which will afford plenty of
+     * precision left over after the division anyway.
+     */
+    uint64_t  b = rxfc->rwm - rxfc->esrwm;
+    OSSL_TIME now, dt, t_window;
+
+    if (b == 0)
+        return 0;
+
+    now      = rxfc->now(rxfc->now_arg);
+    dt       = ossl_time_subtract(now, rxfc->epoch_start);
+    t_window = ossl_time_divide(ossl_time_multiply(dt, rxfc->cur_window_size), b);
+
+    return ossl_time_compare(t_window, ossl_time_multiply(rtt, 4)) < 0;
+}
+
+static void rxfc_adjust_window_size(QUIC_RXFC *rxfc, uint64_t min_window_size,
+                                    OSSL_TIME rtt)
+{
+    /* Are we sending updates too often? */
+    uint64_t new_window_size;
+
+    new_window_size = rxfc->cur_window_size;
+
+    if (rxfc_should_bump_window_size(rxfc, rtt))
+        new_window_size *= 2;
+
+    if (new_window_size < min_window_size)
```

Comment:
Should it be ossl_assert() condition type? Or can this happen in normal operation? Hmm... I suppose it can happen on the connection level rxfc.